### PR TITLE
expat: use cmake configuration file

### PR DESCRIPTION
**Pull request by neheb**
Signed-off-by: Rosen Penev <rosenp@gmail.com>

### CODE REVIEW

**Comment by neheb** [comment]
```
  #cmakedefine XML_DEV_URANDOM
```

meson fails to substitute it. Why do I feel I reported this before?

**Comment by neheb** [comment]
@guijan how is libbsd-overlay supposed to work? I'm getting.

```
ccache cc -Isubprojects/expat-2.4.8/libexpat.so.1.6.11.p -Isubprojects/expat-2.4.8 -I../subprojects/expat-2.4.8 -I../subprojects/expat-2.4.8/lib -fvisibility=hidden -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -O0 -g -DXML_ENABLE_VISIBILITY=1 -DHAVE_EXPAT_CONFIG_H -fno-strict-aliasing -DXML_DEV_URANDOM -fPIC -isystem /usr/include/bsd -DLIBBSD_OVERLAY -MD -MQ subprojects/expat-2.4.8/libexpat.so.1.6.11.p/lib_xmlparse.c.o -MF subprojects/expat-2.4.8/libexpat.so.1.6.11.p/lib_xmlparse.c.o.d -o subprojects/expat-2.4.8/libexpat.so.1.6.11.p/lib_xmlparse.c.o -c ../subprojects/expat-2.4.8/lib/xmlparse.c
In file included from ../subprojects/expat-2.4.8/lib/xmlparse.c:116:
/usr/include/bsd/stdlib.h:30:25: error: no include path in which to search for stdlib.h
   30 | #include_next <stdlib.h>
```

**Comment by guijan** [comment]
When switching from libbsd to libbsd-overlay, remove all the lines that include headers from the `bsd/` subdirectory in the system include path, and include the equivalent standard headers if they're not included already.

In expat's case, these lines need to be removed: 
https://github.com/libexpat/libexpat/blob/3bab6c09bbe8bf42d84b81563ddbcf4cca4be838/expat/lib/xmlparse.c#L114-L117

I tested compiling expat tag R_2_4_8 with this PR's current meson.build and meson_options.txt with the change above, use_libbsd enabled, and the `#cmakedefine` fix on Hyperbola Linux and it succeeded.

**Comment by guijan** [comment]
Yes, that would also work.

**Comment by neheb** [comment]
sounds like another PR for upstream :).

**Comment by eli-schwartz** [comment]
Huh, why?

**Comment by neheb** [comment]
to avoid this libbsd mess.

**Comment by neheb** [comment]
@eli-schwartz just read your earlier comment.

I don't think cc.has_function works without adding a dependency to libbsd. Same with has_header_symbol, which uses regular system includes instead of the libbsd overlay.

Actually speaking of the former, it can't be used with libobsd as a subproject. Need to switch to the latter.

**Comment by eli-schwartz** [comment]
But the point is that HAVE_LIBBSD says "the bsd functions are in a namespaced header space".
HAVE_FUNCTIONNAME says "I have the functions".

If you're using an overlay, then the former is false. libbsd itself also provides an overlay mechanism, and offers a choice.

**Comment by neheb** [comment]
libobsd (available as a wrap here) is overlay only.

On that note, even has_header_symbol throws

```
subprojects/expat-2.4.8/meson.build:43:7: ERROR: Dependencies must be external dependencies
```

when using libobsd as a subproject. strange.

**Comment by neheb** [comment]
I took out the cmake config commit. Now it's just a cleanup one.

**Comment by neheb** [comment]
@xclaesse rebased.

File: ci_config.json [hunk 1]
```diff
@@ -22,7 +22,6 @@
     "termbox",
     "unit-system",
     "zstd",
-    "libexif",
```
Comment (xclaesse) [inline ci_config.json:25]: This PR is about expat, no?

File: subprojects/packagefiles/expat/meson.build [hunk 2]
```diff
@@ -153,23 +130,16 @@ _files = [
...
-  extra_kwargs += {'vs_module_defs' : 'lib/libexpat.def'}
-endif
-
 libexpat = library(
   'expat',
   _files,
   include_directories : include_directories('lib'),
   dependencies : dep_libbsd,
-  version : '1.6.11',
+  version : '1.8.8',
```
Comment (xclaesse) [inline subprojects/packagefiles/expat/meson.build:138]: Better define that at the top of the file, so we are less likely to forget when bumping version in project().

Comment (neheb) [reply subprojects/packagefiles/expat/meson.build:138]: Better solution would be to derive it from the project version. Not too sure how to do so. @eli-schwartz knows more I think.

Comment (eli-schwartz) [reply subprojects/packagefiles/expat/meson.build:138]: The library version is libtool `9:8:8`, which is mathematically expressed as 1.8.8, it has no basis in the project version.

Upstream just uses code comments to remember to bump it in every file whenever editing it in one.

Comment (neheb) [reply subprojects/packagefiles/expat/meson.build]: This was fixed.

File: ci_config.json (hunk 1 above)
Comment (neheb) [reply ci_config.json]: Fixed.

**Comment by neheb** [comment]
anything else?

**Comment by xclaesse** [comment]
Fine with me

File: subprojects/packagefiles/expat/meson.build [hunk 3]
```diff
@@ -24,44 +24,27 @@ project(
...
-endif
-
-if cc.has_header_symbol(header, 'arc4random_buf')
-  config.set('HAVE_ARC4RANDOM_BUF', true)
-elif cc.has_header_symbol(header, 'arc4random')
-  config.set('HAVE_ARC4RANDOM', true)
-endif
+dep_libbsd = dependency('libbsd-overlay', required : get_option('use_libbsd'), allow_fallback: false)
+config.set('HAVE_ARC4RANDOM_BUF', cc.has_header_symbol('stdlib.h', 'arc4random_buf', dependencies: dep_libbsd))
+config.set('HAVE_ARC4RANDOM', cc.has_header_symbol('stdlib.h', 'arc4random', dependencies: dep_libbsd))
```
Comment (eli-schwartz) [inline subprojects/packagefiles/expat/meson.build:40]: Can't we just allow the dependency fallback, but if we do fallback then check if dep_libbsd.type_name() == 'internal' and unconditionally set these without running has_header_symbol at all?

**Review by eli-schwartz — CHANGES_REQUESTED** [review]

File: subprojects/packagefiles/expat/meson.build (hunk 3 above)
Comment (neheb) [reply subprojects/packagefiles/expat/meson.build:40]: sounds like quite the hack :)

Comment (neheb) [reply subprojects/packagefiles/expat/meson.build]: Added this suggestion.

**Pull request merged**